presto-main/src/main/java/com/facebook/presto/sql/planner/plan/LateralJoinNode.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.plan;
+
+import com.facebook.presto.sql.planner.Symbol;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * For every row from {@link #input} a {@link #subquery} relation is calculated.
+ * Then input row is cross joined with subquery relation and returned as a result.
+ *
+ * INNER - does not return any row for input row when subquery relation is empty
+ * LEFT - does return input completed with NULL values when subquery relation is empty
+ */
+@Immutable
+public class LateralJoinNode
```

Comment:
We were thinking about that with @kokosing and we have decided that the simplest, most direct names would be the best. `LateralJoin` and `Apply` are well understood concepts in SQL world. Therefore instead of naming them somehow different, we decided to name them what they are.
After all, one can say that subqueries or `Apply` is also a SQL syntactic sugar since it is possible to rewrite them to more basic nodes.